File: app/src/main/java/org/eyeseetea/malariacare/views/filters/OrgUnitProgramFilterView.java
package org.eyeseetea.malariacare.views.filters;

import android.content.Context;
import android.util.AttributeSet;
import android.view.View;
import android.widget.AdapterView;
import android.widget.FrameLayout;

import org.eyeseetea.malariacare.R;
import org.eyeseetea.malariacare.data.database.model.OrgUnitDB;
import org.eyeseetea.malariacare.data.database.model.ProgramDB;
import org.eyeseetea.malariacare.data.database.utils.PreferencesState;
import org.eyeseetea.malariacare.layout.adapters.filters.FilterOrgUnitArrayAdapter;
import org.eyeseetea.malariacare.layout.adapters.filters.FilterProgramArrayAdapter;
import org.eyeseetea.malariacare.presentation.presenters.OrgUnitProgramFilterPresenter;
import org.eyeseetea.malariacare.views.CustomSpinner;

import java.util.List;

public class OrgUnitProgramFilterView
        extends FrameLayout implements OrgUnitProgramFilterPresenter.View {

    public enum FilterType {EXCLUSIVE,NON_EXCLUSIVE}

    public interface FilterChangedListener {
        void onProgramFilterChanged(ProgramDB programFilter);
        void onOrgUnitFilterChanged(OrgUnitDB orgUnitFilter);
    }

    private FilterChangedListener mFilterChangedListener;

    private CustomSpinner orgUnitFilterSpinner;
    private CustomSpinner programFilterSpinner;

    private OrgUnitProgramFilterPresenter presenter;


    public OrgUnitProgramFilterView(Context context, AttributeSet attributeSet) {
        super(context, attributeSet);
        init(context, attributeSet);
    }

    public void setFilterChangedListener(FilterChangedListener filterChangedListener) {
        mFilterChangedListener = filterChangedListener;
    }

    private void init(Context context, AttributeSet attributeSet) {
        inflate(context, R.layout.view_orgunit_program_filter, this);

        initializeOrgUnitViews();
        initializeProgramViews();
        initializePresenter();
    }

    private void initializePresenter() {
        presenter = new OrgUnitProgramFilterPresenter();
        presenter.attachView(this);

    }

    private void initializeOrgUnitViews() {
        orgUnitFilterSpinner = (CustomSpinner) findViewById(R.id.spinner_orgUnit_filter);

        orgUnitFilterSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {

            @Override
            public void onItemSelected(AdapterView<?> parent, View view,
                    int position, long id) {
                OrgUnitDB orgUnit = (OrgUnitDB) parent.getItemAtPosition(position);

                presenter.onOrgUnitSelected(orgUnit);
            }

            @Override
            public void onNothingSelected(AdapterView<?> arg0) {

            }
        });
    }

    private void initializeProgramViews() {
        programFilterSpinner = (CustomSpinner) findViewById(R.id.spinner_program_filter);

        programFilterSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {

            @Override
            public void onItemSelected(AdapterView<?> parent, View view,
                    int position, long id) {
                ProgramDB program = (ProgramDB) parent.getItemAtPosition(position);

                presenter.onProgramSelected(program);
            }

            @Override
            public void onNothingSelected(AdapterView<?> adapterView) {

            }
        });
    }

    @Override
    public void renderPrograms(List<ProgramDB> programs) {
        programFilterSpinner.setAdapter(new FilterProgramArrayAdapter(getContext(), programs));
    }

    @Override
    public void renderOrgUnits(List<OrgUnitDB> orgUnits) {
        orgUnitFilterSpinner.setAdapter(new FilterOrgUnitArrayAdapter(getContext(), orgUnits));
    }

    @Override
    public void notifyProgramFilterChange(ProgramDB programFilter) {
        if (mFilterChangedListener != null) {
            mFilterChangedListener.onProgramFilterChanged(programFilter);
        }
    }

    @Override
    public void notifyOrgUnitFilterChange(OrgUnitDB orgUnitFilter) {
        if (mFilterChangedListener != null) {
            mFilterChangedListener.onOrgUnitFilterChanged(orgUnitFilter);
        }
    }

    @Override
    public void unSelectOrgUnitFilter() {
        orgUnitFilterSpinner.setSelection(0,true,true);
    }

    @Override
    public void unSelectProgramFilter() {
        programFilterSpinner.setSelection(0,true,true);
    }

    @Override
    public void selectOrgUnitFilter(int indexToSelect) {
        orgUnitFilterSpinner.setSelection(indexToSelect,true,true);
    }

    @Override
    public void selectProgramFilter(int indexToSelect) {
        programFilterSpinner.setSelection(indexToSelect,true,true);
    }

    public ProgramDB getSelectedProgramFilter() {
        return presenter.getSelectedProgramFilter();
    }

    public OrgUnitDB getSelectedOrgUnitFilter() {
        return presenter.getSelectedOrgUnitFilter();
    }


    public void setFilterType(FilterType filterType){
        if (filterType == FilterType.EXCLUSIVE)
            presenter.setExclusiveFilter(true);
        else
            presenter.setExclusiveFilter(false);
    }

    public void changeSelectedFilters(String programUidFilter, String orgUnitUidFilter) {
        presenter.changeSelectedFilters(programUidFilter,orgUnitUidFilter);
    }
}

File: app/src/main/java/org/eyeseetea/malariacare/views/filters/OrgUnitProgramFilterView.kt
package org.eyeseetea.malariacare.views.filters

import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.FrameLayout
import kotlinx.android.synthetic.main.view_orgunit_program_filter.view.*
import org.eyeseetea.malariacare.R
import org.eyeseetea.malariacare.factories.MetadataFactory
import org.eyeseetea.malariacare.presentation.executors.WrapperExecutor
import org.eyeseetea.malariacare.presentation.presenters.OrgUnitProgramFilterPresenter

class OrgUnitProgramFilterView(context: Context, attributeSet: AttributeSet) :
    FrameLayout(context, attributeSet), OrgUnitProgramFilterPresenter.View {

    // TODO: When all callers be converted to kotlin change this interface by functions
    interface FilterChangedListener {
        fun onProgramFilterChanged(programFilter: String)
        fun onOrgUnitFilterChanged(orgUnitFilter: String)
    }

    enum class FilterType {
        EXCLUSIVE, NON_EXCLUSIVE
    }

    private var mFilterChangedListener: FilterChangedListener? = null

    private lateinit var presenter: OrgUnitProgramFilterPresenter

    val selectedProgramFilter: String
        get() = presenter.selectedUidProgram

    val selectedOrgUnitFilter: String
        get() = presenter.selectedUidOrgUnit

    init {
        init(context)
    }

    fun setFilterChangedListener(filterChangedListener: FilterChangedListener) {
        mFilterChangedListener = filterChangedListener
    }

    fun setFilterType(filterType: FilterType) {
        if (filterType == FilterType.EXCLUSIVE)
            presenter.setExclusiveFilter(true)
        else
            presenter.setExclusiveFilter(false)
    }

    fun changeSelectedFilters(programUidFilter: String, orgUnitUidFilter: String) {
        presenter.changeSelectedFilters(programUidFilter, orgUnitUidFilter)
    }

    override fun showPrograms(programNames: List<String>) {
        spinner_program_filter.adapter =
            ArrayAdapter(context, R.layout.simple_spinner_item, programNames)
    }

    override fun showOrgUnits(orgUnitNames: List<String>) {
        spinner_orgUnit_filter.adapter =
            ArrayAdapter(context, R.layout.simple_spinner_item, orgUnitNames)
    }

    override fun notifyProgramFilterChange(programFilter: String) {
        mFilterChangedListener?.onProgramFilterChanged(programFilter)
    }

    override fun notifyOrgUnitFilterChange(orgUnitFilter: String) {
        mFilterChangedListener?.onOrgUnitFilterChanged(orgUnitFilter)
    }

    override fun unSelectOrgUnitFilter() {
        spinner_orgUnit_filter.setSelection(0, true, true)
    }

    override fun unSelectProgramFilter() {
        spinner_program_filter.setSelection(0, true, true)
    }

    override fun selectOrgUnitFilter(indexToSelect: Int) {
        spinner_orgUnit_filter.setSelection(indexToSelect, true, true)
    }

    override fun selectProgramFilter(indexToSelect: Int) {
        spinner_program_filter!!.setSelection(indexToSelect, true, true)
    }

    private fun init(context: Context) {
        View.inflate(context, R.layout.view_orgunit_program_filter, this)

        initializeOrgUnitViews()
        initializeProgramViews()
        initializePresenter()
    }

    private fun initializePresenter() {
        presenter = OrgUnitProgramFilterPresenter(
            WrapperExecutor(),
            MetadataFactory.provideGetOrgUnitsUseCase(),
            MetadataFactory.provideGetProgramsUseCase()
        )

        presenter.attachView(
            this,
            context.getString(R.string.filter_all_org_units),
            context.getString(R.string.filter_all_org_assessments)
        )
    }

    private fun initializeOrgUnitViews() {
        this.spinner_orgUnit_filter.onItemSelectedListener =
            object : AdapterView.OnItemSelectedListener {

                override fun onItemSelected(
                    parent: AdapterView<*>,
                    view: View,
                    position: Int,
                    id: Long
                ) {
                    val orgUnitName = parent.getItemAtPosition(position) as String

                    presenter.onOrgUnitSelected(orgUnitName)
                }

                override fun onNothingSelected(arg0: AdapterView<*>) {
                }
            }
    }

    private fun initializeProgramViews() {
        this.spinner_program_filter.onItemSelectedListener =
            object : AdapterView.OnItemSelectedListener {

                override fun onItemSelected(
                    parent: AdapterView<*>,
                    view: View?,
                    position: Int,
                    id: Long
                ) {
                    val programName = parent.getItemAtPosition(position) as String

                    presenter.onProgramSelected(programName)
                }

                override fun onNothingSelected(adapterView: AdapterView<*>) {
                }
            }
    }
}
